recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,376 @@
+
+### Environment variables
+In a significant number of providers, as per documentation, environment variables are used one of the methods of saving sensitive credential data along with insensitive data. We allow the users to set environment variables for provider configuration. For sensitive information, we recommend the users save these values as secrets and point to them inside the env block.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+     providers: [...]
+     env: {
+        'MY_ENV_VAR_1': 'my_value'
+        secrets: {                       // Individual Secrets from SecretStore
+        'MY_ENV_VAR_2': {
+            source: secretStoreConfig.id
+            key: 'envsecret.one'
+          }
+        'MY_ENV_VAR_3': {
+            source: secretStoreConfig.id
+            key: 'envsecret.two'
+          }
+        }
+      }
+```
+
+Environment variables apply to all providers configured in the environment. The system cannot set two separate values for the same environment variable for multiple provider configurations. In such cases, per provider documentation, there may be alternatives that users can avail (eg. For GCP, users can set  *credentials* field inside each instance of provider config as opposed to using env variable GOOGLE_APPLICATION_CREDENTIALS).
+
+
+The system will allow for ability to set up multiple configurations for the same provider using the keyword *alias*. Validation for configuration 'correctness' will be handled by Terraform with calls to terraform init and terraform apply.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+      providers: [
+      {
+        name: 'azurerm',
+        properties: {
+          subscriptionid: 1234,
+          secrets: {                  // Individual Secrets from SecretStore
+            'my_secret_1': {
+              source: secretStoreAz.id
+              key: 'secret.one'
+            }
+            'my_secret_2': {
+               source: secretStoreAzPayment.id
+              key: 'secret.two'
+            }
+          }
+        }
+      },
+      {
+        name: 'azurerm',
+        properties: {
+          subscriptionid: 1234,
+          alias: 'az-paymentservice'
+       }
+     }]
+...       
+```
+Configuration for providers as described in this document will take precedence over provider credentials stored in UCP (currently these include azurerm, aws, kubernetes providers). So, for eg. In the scenario where credentials for Azure are saved with UCP during Radius install and a Terraform recipe created by an operator declares 'azurerm' under the the *required_providers* block; If there exists a provider configuration under the *providers* block under *recipeConfig*, these would take precedence and used to build the Terraform configuration file instead of Azure credentials stored in UCP. 
+
+
+### Example Bicep Input :
+``` diff
```

Comment:
@lakshmimsft - can you update this to be reflect the option we chose in the meeting? Looks like it's describing both options but we made a decision.

Reply:
Oh, I see you did that below.